operator/controllers/resources/role_controller.go

```diff
@@ -0,0 +1,108 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gravitational/teleport/api/types"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gravitational/teleport/api/client"
+	resourcesv5 "github.com/gravitational/teleport/operator/apis/resources/v5"
+	"github.com/gravitational/trace"
+)
+
+// RoleReconciler reconciles a TeleportRole object
+type RoleReconciler struct {
+	kclient.Client
+	Scheme         *runtime.Scheme
+	TeleportClient *client.Client
+}
+
+//+kubebuilder:rbac:groups=resources.teleport.dev,resources=roles,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=resources.teleport.dev,resources=roles/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=resources.teleport.dev,resources=roles/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the TeleportRole object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *RoleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return ResourceBaseReconciler{
+		Client:         r.Client,
+		DeleteExternal: r.Delete,
+		UpsertExternal: r.Upsert,
+	}.Do(ctx, req, &resourcesv5.TeleportRole{})
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *RoleReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&resourcesv5.TeleportRole{}).
+		Complete(r)
+}
+
+func (r *RoleReconciler) Delete(ctx context.Context, obj kclient.Object) error {
+	return r.TeleportClient.DeleteRole(ctx, obj.GetName())
+}
+
+func (r *RoleReconciler) Upsert(ctx context.Context, obj kclient.Object) error {
+	k8sResource, ok := obj.(*resourcesv5.TeleportRole)
+	if !ok {
+		return fmt.Errorf("failed to convert Object into resource object: %T", obj)
+	}
+	teleportResource := k8sResource.ToTeleport()
+
+	existingResource, err := r.TeleportClient.GetRole(ctx, teleportResource.GetName())
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+
+	newCondition, ownershipErr := checkOwnership(existingResource)
+	// Setting the condition before returning a potential ownership error
+	meta.SetStatusCondition(&k8sResource.Status.Conditions, newCondition)
+	if err := r.Status().Update(ctx, k8sResource); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if ownershipErr != nil {
+		return trace.Wrap(ownershipErr)
```

Comment:
If not owned by the operator and  we return an error, won't it requeue the operation forever?

Reply:
Yes, but I'm not sure if the alternative of not caring anymore is saner. If the user edits the resource to allow adoption it won't be reconciled. Maybe this could be handled with another status.condition.